src/Tokenizer.php

```diff
@@ -847,222 +839,138 @@ private function makeRegexFromList(array $values, bool $sorted = false): string
-        if (($firstChar === '@' || $firstChar === ':') && $secondChar !== '') {
-            $value = null;
-            $type  = Token::TOKEN_TYPE_VARIABLE;
-
-            // If the variable name is quoted
-            if ($secondChar === '"' || $secondChar === '\'' || $secondChar === '`') {
-                $value = $firstChar . $this->getNextQuotedString($string, $offset + 1);
-            } else {
-                // Non-quoted variable name
-                preg_match('/\G[@:][\w.$]+/', $string, $matches, 0, $offset);
-                if ($matches) {
-                    $value = $matches[0];
-                }
-            }
-
-            if ($value !== null) {
-                return new Token($type, $value);
-            }
-        }
-
-        // Number (decimal, binary, or hex)
-        if (
-            preg_match(
-                $this->nextTokenRegexNumber,
-                $string,
-                $matches,
-                0,
-                $offset,
-            )
-        ) {
-            return new Token(Token::TOKEN_TYPE_NUMBER, $matches[0]);
-        }
-
-        // Boundary Character (punctuation and symbols)
-        if (preg_match($this->nextTokenRegexBoundaryCharacter, $string, $matches, 0, $offset)) {
-            return new Token(Token::TOKEN_TYPE_BOUNDARY, $matches[0]);
-        }
+        // remap named groups to unnamed
+        $namedGroupsRegex        = '(?:' . implode('|', array_map(static fn ($v) => preg_quote($v, '~'), $regexIndexToName)) . ')';
+        $regexWithoutNamedGroups = preg_replace('~(?<=\()\?<' . $namedGroupsRegex . '>~', '', $regex);
+        assert($regexWithoutNamedGroups !== null);
+        $regexWithoutNamedGroups = preg_replace_callback(
+            '~(?<=\(\?)&' . $namedGroupsRegex . '(?=\))~',
+            static fn ($matches) => array_flip($regexIndexToName)[substr($matches[0], 1)],
+            $regexWithoutNamedGroups,
+        );
+        assert($regexWithoutNamedGroups !== null);
+
+        return [$regexWithoutNamedGroups, $regexIndexToName];
+    }
 
-        // A reserved word cannot be preceded by a '.'
-        // this makes it so in "mytable.from", "from" is not considered a reserved word
-        if ($previous === null || $previous->value() !== '.') {
-            // Top Level Reserved Word
-            if (
-                preg_match(
-                    $this->nextTokenRegexReservedToplevel,
-                    $upper,
-                    $matches,
-                    0,
-                    $offset,
-                )
-            ) {
-                return new Token(
-                    Token::TOKEN_TYPE_RESERVED_TOPLEVEL,
-                    substr($string, $offset, strlen($matches[0])),
-                );
-            }
+    /**
+     * Takes a SQL string and breaks it into tokens.
+     * Each token is an associative array with type and value.
+     *
+     * @param string $string The SQL string
+     */
+    public function tokenize(string $string): Cursor
```

Comment:
Can you reorganize the code so as to minimize the diff? For instance, you moved `tokenize` down in the file, was there a reason for that? It makes the diff bigger than it needs to be IMO.

Reply:
I have reverted the performance optimization of https://github.com/php/php-src/issues/14423 for PHP 8.3 and below. The code should be now as simplest as possible.

Reply:
@greg0ire can you please do a final review?